deid/dicom/groups.py

```diff
@@ -0,0 +1,108 @@
+"""
+
+groups: functions to derive groups of fields or values
+
+Copyright (c) 2020 Vanessa Sochat
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+"""
+
+
+from deid.logger import bot
+from .tags import remove_sequences
+from .fields import get_fields, expand_field_expression
+
+import os
+
+
+def extract_values_list(dicom, actions):
+    """Given a list of actions for a named group (a list) extract values from
+       the dicom based on the list of actions provided. This function
+       always returns a list intended to update some lookup to be used
+       to further process the dicom.
+    """
+    values = []
+    fields = get_fields(dicom)
+    for action in actions:
+
+        # Just grab the entire value string for a field, no parsing
+        if action["action"] == "FIELD":
+            subset = expand_field_expression(
+                field=action["field"], dicom=dicom, contenders=fields
+            )
+            [values.append(dicom.get(field)) for field in subset]
```

Comment:
A few questions:

 1. How do we check for DICOM OB datatype (what would that look like given the above?)
 2. `fields = get_fields(dicom)` does not include any fields with None values - does this address your concern?
 3. What is an example real world case for wanting this?

Reply:
> 1. How do we check for DICOM OB datatype (what would that look like given the above?)

I need to dig into this a bit more but I think this is going to have to wait for #118.  pydicom Dataset.get() returns a different object based on what is used for the key.  Since we're using the tag names, the tag value is returned. When we're in a position where we can call get with actual tag for the element, get() returns a DataElement instance.  Once we have a DataElement instance, we can interrogate the VR property to determine if the value is a OB (OD, OF, and OW should also be included with OB).

> 2. ```fields = get_fields(dicom)``` does not include any fields with None values - does this address your concern?

Oops.  I didn't realize what I actually did here.  The None value is added when the ```%values``` group references a field that doesn't exist within the dicom header at all:
```
%fields badsection
FIELD DoesNotExist
FIELD FabricatedName
```

> 3. What is an example real world case for wanting this?

This might be a stretch, but...
```
# Sample Recipe:
%values scanner_info
FIELD Manufacturer
FIELD DeviceSerialNumber

# Sample Header 1:
(0008,0070) : LO   Len: 8      Manufacturer                   Value: [SIEMENS ]
(0018,1000) : LO   Len: 6      DeviceSerialNumber             Value: [12345 ]

# Sample Header 2:
(0008,0070) : LO   Len: 8      Manufacturer                   Value: [AGFA ]
(0018,1000) : LO   Len: 6      DeviceSerialNumber             Value: [983 ]

```

I could foresee wanting to avoid removing '983' - much in the same way we would want to not include 'J' from 'SIMPSON^HOMER^J^^' in our patient name SPLIT example.

Reply:
1. Gotcha! I saw the same with respect to indexing by a tag. Do you think it's worth bumping the pydicom version (note that right now I set an upper limit) so we are developing #118 for latest?
2. That looks like a bug - we shouldn't add a None if the field isn't present. Let me fix this, does this make sense to do? `[values.add(dicom.get(field)) for field in subset if field in dicom]`
3. I understand the use case - I'd advocate that we don't add this extra feature until we are absolutely sure there is a case in the real world.

Reply:
1. Glancing through some of the commits, it looks like the releases since pydicom 1.2.1 have some enhancements that may help with #118.  I'd vote for going for it and developing #118 for latest.
2. Makes sense to me.
3. That's fine.

Reply:
okay sounds good.